README
代码下载于 http://www.man7.org/tlpi/code/index.html

相关说明可参考以上链接

make之后,没有出现报错,就可以测试和验证了.



Gidday!

This is the code for the book "The Linux Programming Interface"
and this is a note from me, the author, Michael Kerrisk.

For instructions on building the programs, see the file BUILDING.

For notes on changes that have been made to the code since it was
published in the book, see the file CHANGES.


Source code licensing
=====================

All complete programs provided in this distribution are covered by
the GNU General Public License (Version 3), a copy of which is
contained in the file COPYING.gpl-v3, which should have arrived with
this tarball.  The library functions (in the lib/ directory) are
covered by the GNU Lesser General Public License (Version 3); see the
file COPYING.lgpl-v3 provided with this tarball.


A note on the source code
=========================

The source code is available in two versions: "dist" and "book".
The "book" version contains the program source files as published in
the book. The source files in the "dist" version contain extra code
beyond that published in the book. The differences between the "dist"
and "book" versions are as follows:

a) The "dist" versions of some programs contain extra comments.
   These additional comments were stripped out of the printed version
   to make the published versions of the programs shorter. (The book
   itself contains text describing the operation of the programs.)

b) In a few cases, some changes have been incorporated into the
   "dist" versions to make it possible to compile programs on UNIX
   implementations other than Linux, so that you can try them out
   on other implementations if you wish.  Where necessary, the
   additional code is conditionally compiled using the following
   platform-specific macros:

        __linux__       Defined on Linux
        __sun           Defined on Solaris
        __FreeBSD__     Defined on FreeBSD
        __NetBSD__      Defined on NetBSD
        __OpenBSD__     Defined on OpenBSD
        __APPLE__       Defined on Mac OS X
        __hpux          Defined on HP-UX
        __osf__         Defined on Tru64 UNIX (formerly DEC OSF1)
        __sgi           Defined on Irix
        _AIX            Defined on AIX

c) In the "dist" version, some programs have extra functionality beyond
   that in the "book" versions. Where this is significant, comments in
   the programs explain the differences.


Subdirectories
==============

Under the 'tlpi' directory are a number of subdirectories. Each
subdirectory corresponds to one or more chapters of the book.
The following paragraphs give brief notes on the contents of
each subdirectory.

Note that in some cases, files are (hard) linked to appear in more than
one directory. This is particularly the case for each of the files in
the 'lib' directory, most of which are also linked in the directory
of the chapter relating to that file.

Directory       Files for Chapter...

lib             This contains library routines used by other 
                programs. The tlpi_hdr.h and error_functions.* 
                files are located here.
                
progconc        3 (System Programming Concepts)

fileio          4 and 5 (File I/O)

proc            6 (Processes)

memalloc        7 (Memory Allocation)

users_groups    8 (Users and Groups)

proccred        9 (Process Credentials)

time            10 (Time)

syslim          11 (System Limits and Options)

sysinfo         12 (System and Process Information)

filebuff        13 (File I/O Buffering)

filesys         14 (File Systems)

files           15 (File Attributes)

xattr           16 (Extended Attributes)

acl             17 (Access Control Lists)

dirs_links      18 (Directories and Links)

inotify         19 (Monitoring File Events)

signals         20 to 22 (Signals)

timers          23 (Timers and Sleeping)

procexec        24 (Process Creation), 25 (Process Termination),
                26 (Monitoring Child Processes), 27 (Program Execution),
                and 28 (Further Details on Process Creation and Program
                Execution)

threads         29 to 33 (POSIX Threads)

pgsjc           34 (Process Groups, Sessions, and Job Control)

procpri         35 (Process Priorities and Scheduling)

procres         36 (Process Resources)

daemons         37 (Daemons)

cap             39 (Capabilities)

loginacct       40 (Login Accounting)

shlibs          41 and 42 (Shared Libraries)
    
pipes           44 (Pipes and FIFOs)

svipc           45 (System V IPC)
    
svmsg           46 (System V Message Queues)

svsem           47 (System V Semaphores)

svshm           48 (System V Shared Memory)

mmap            49 (Memory Mappings)

vmem            50 (Virtual Memory Operations)

pmsg            52 (POSIX Message Queues)

psem            53 (POSIX Semaphores)

pshm            54 (POSIX Shared Memory)

filelock        55 (File Locking)

sockets         56 to 61 (Sockets and Network Programming)

tty             62 (Terminals)

altio           63 (Alternative I/O Models)

pty             64 (Pseudoterminals)

getopt          Appendix B: Parsing Command-Line Options

In addition, the following supplementary code is included (relating
to topics NOT covered in TLPI):

namespaces      Code examples for namespaces, mainly related to my LWN.net
                article series starting at https://lwn.net/Articles/531114/


CHANGES
This file summarizes the changes that have been made since publication
to the program examples printed in "The Linux Programming Interface".
Background on some of these changes can be found in the online errata
for TLPI, available at http://man7.org/tlpi/errata/.

2010-11-13
        sockets/us_abstract_bind.c
                The code was improved as per the errata for page 1176.

2011-01-17
        timers/real_timer.c
                A mistake in the ordering of the code in main() was fixed.
                See the erratum for page 483.

2011-02-17
        psem/thread_incr_psem.c
                Fixed an error in a comment. See the erratum for page 1102.

2011-04-05
        threads/thread_multijoin.c.
                Fixed a race condition. See the erratum for page 649.
        threads/prod_condvar.c
                Fixed a race condition. The problem was similar to that
                described in the erratum for page 649.
        threads/prod_no_condvar.c
                Fixed a race condition. The problem was similar to that
                described in the erratum for page 649.

2011-04-19
        altio/demo_sigio.c
                Fixed a race condition. See the erratum for page 1349.
        daemons/daemon_SIGHUP.c
                Fixed a race condition. See the erratum for page 774.

2011-05-18
        signals/t_kill.c
                Fixed an error in a diagnostic message. See the erratum
                for page 406.

2011-07-06
        Makefile
                Added missing "memalloc" to the directory list.
        memalloc/free_and_sbrk.c
                Added feature test macro definition (_BSD_SOURCE).
                See the erratum for page 142.

2011-08-11
        acl/acl_view.c
                Fixed a small bug as per the erratum for page 336.

2011-09-04
        pipes/popen_glob.c
        procexec/make_zombie.c
        sockets/inet_sockets.c
                Removed an unnecessary assignment statement that added
                a terminating null byte to the string buffer output by
                snprintf().  See the erratum for page 555.

2011-12-05
        dirs_links/t_unlink.c
                Added a comment referring to the erratum for page 348.

2011-12-06
        pty/unbuffer.c
                Change parent exit status when read() returns <= 0
                from EXIT_FAILURE to EXIT_SUCCESS.

2011-12-13
        signals/catch_rtsigs.c
                Restore 3 lines that were accidentally omitted at the
                end of this program. See the erratum for page 463.

2011-12-31
        threads/thread_incr.c
                Make global variable 'glob' volatile, so that the program
                more easily produces "incorrect" behavior, even in the face
                of compiler optimizations. See the erratum for page 632.

2012-02-16
        lib/alt_functions.c
                Fix indentation in ALT_posix_openpt().

2012-04-03
        lib/Makefile
        lib/Build_ename.sh
                Refactor, so that Build_ename.sh produces output on stdout,
                rather than to a named file.

2012-04-06
        lib/itimerspec_from_str.c
                Conditionally make the content of this file empty if
                compiling on MacOSX, since that operating system doesn't
                define the 'itimerspec' structure.
                (Only in "dist" version of code.)
        Makefile.inc.MacOSX
                Remove '-lrt' from the IMPL_LDLIBS definition, since there
                is no librt on MacOSX.

2012-05-04
        fileio/seek_io.c
                Fix a typo in comment at top of the program.
                (Only in "dist" version of code.)

2012-05-11
        psem/Makefile
                Correct the makefile to use "cc -pthreads" for POSIX
                semaphores. Formerly, the makefile used "cc -lrt", but
                recent toolchain changes mean that that this no longer works
                ("cc -pthreads" always worked.) See the erratum for page 1061.
                (Thanks to Robert P. J. Day.)
        pshm/Makefile
                Correct a bug in the makefile that caused link errors.
                (Formerly, the makefile worked, but this was fortuitous;
                recent toolchain changes revealed the bug.)

2012-05-13
        pshm/README
                Fix a wordo.

2012-05-20
        README
                Various small fixes.
                (Thanks to Robert P. J. Day.)

2012-05-25
        BUILDING
                Various small fixes.
                (Thanks to Robert P. J. Day.)

2012-05-26
        memalloc/free_and_sbrk.c
                Fix an error in the code comments at the top of the program.
                (Only in "dist" version of code.)
                (Thanks to Robert P. J. Day.)

2012-05-31
        svmsg/svmsg_info.c
        svsem/svsem_info.c
        svshm/svshm_info.c
                Eliminate unnecessary inclusion of <sys/ipc.h> header file.
                (Thanks to Robert P. J. Day.)

2012-06-05
        COPYING
                Renamed to COPYING.agpl-v3
        Added new files:
                COPYING.gpl-v3
                        Copy of the GNU General Public License, version 3
                COPYING.lgpl-v3
                        Copy of the GNU Lesser General Public License, version 3
        lib/*
                Changed the license of the library functions in the /lib
                directory to GNU Lesser General Public License, version 3.
        sockets/read_line_buf.c
        sockets/read_line_buf.h
                Created links for the sockets/read_line_buf.{c,h} files in the
                lib/ directory, so that these files are licensed LGPLv3.

        README
                Updated to note the licensing of the library functions.

2012-07-05
        filesys/t_statfs.c
        filesys/t_statvfs.c
                Changed several printf() statements to use unsigned types for
                various fields, since (in the case of the 'statvfs' structure at

seccomp/seccomp_arg64.c
/*************************************************************************\
*                  Copyright (C) Michael Kerrisk, 2017.                   *
*                                                                         *
* This program is free software. You may use, modify, and redistribute it *
* under the terms of the GNU General Public License as published by the   *
* Free Software Foundation, either version 3 or (at your option) any      *
* later version. This program is distributed without any warranty.  See   *
* the file COPYING.gpl-v3 for details.                                    *
\*************************************************************************/

/* seccomp_arg64.c

   Some system call arguments can be 64 bytes in size, for example pointers
   and the 'off_t' argument of system calls such as lseek(). The
   'seccomp_data' structure allows for this: the elements of 'args' are each
   64 bits in size.  However, the BBF accumulator register is only 32 bits
   in size.  Therefore, 64-bit system call arguments must be dealt with in
   two pieces.  This program provides a simple example of how to do this.

   The program applies a seccomp filter that causes lseek() calls that
   specify an 'offset' value greater tnan 1000 to fail with an error in
   'errno'. The error number varies, depending on whether or not there
   'offset' argument was bigger than 32 bits.

   See also the kernel source file samples/seccomp/bpf-helper.h
   for some useful helper macros for dealing with 64-bit arguments.
*/
#define _GNU_SOURCE
#include <stddef.h>
#include <fcntl.h>
#include <linux/audit.h>
#include <sys/syscall.h>
#include <linux/filter.h>
#include <linux/seccomp.h>
#include <sys/prctl.h>
#include "tlpi_hdr.h"

static int
seccomp(unsigned int operation, unsigned int flags, void *args)
{
    return syscall(__NR_seccomp, operation, flags, args);
}

static void
install_filter(void)
{
    struct sock_filter filter[] = {

        /* Load architecture */

        BPF_STMT(BPF_LD | BPF_W | BPF_ABS,
                (offsetof(struct seccomp_data, arch))),

        /* Kill process if the architecture is not what we expect */

        BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, AUDIT_ARCH_X86_64, 1, 0),
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_KILL),

        /* Load system call number */

        BPF_STMT(BPF_LD | BPF_W | BPF_ABS,
                 (offsetof(struct seccomp_data, nr))),

        /* Allow system calls other than lseek() */

        BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, __NR_lseek, 1, 0),
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),

        /* Load top 4 bytes of 'offset' argument; fail with errno==2
           if these bytes are nonzero */

        BPF_STMT(BPF_LD | BPF_W | BPF_ABS,
                 (offsetof(struct seccomp_data, args[1]) + sizeof(__u32))),
        BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, 0, 1, 0),
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ERRNO | 2),

        /* Load bottom 4 bytes of 'offset' argument; fail with errno==1
           if the value is > 1000; otherwise allow the system call */

        BPF_STMT(BPF_LD | BPF_W | BPF_ABS,
                 (offsetof(struct seccomp_data, args[1]))),
        BPF_JUMP(BPF_JMP | BPF_JGT | BPF_K, 1000, 0, 1),
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ERRNO | 1),

        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),
    };

    struct sock_fprog prog = {
        .len = (unsigned short) (sizeof(filter) / sizeof(filter[0])),
        .filter = filter,
    };

    if (seccomp(SECCOMP_SET_MODE_FILTER, 0, &prog) == -1)
        errExit("seccomp");
    /* On Linux 3.16 and earlier, we must instead use:
            if (prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, &prog))
                errExit("prctl-PR_SET_SECCOMP");
    */
}

static void
seek_test(int fd, off_t offset)
{
    off_t ret;

    printf("Seek to byte %lld: ", (long long) offset);
    ret = lseek(fd, offset, SEEK_SET);
    if (ret == 0)
        printf("succeeded\n");
    else
        printf("failed with errno = %d\n", errno);
}

int
main(int argc, char **argv)
{
    int fd;

    if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0))
        errExit("prctl");

    install_filter();

    fd = open("/dev/zero", O_RDWR);
    if (fd == -1)
        errExit("open");

    seek_test(fd, 0);
    seek_test(fd, 10000);
    seek_test(fd, 0x100000001);

    exit(EXIT_SUCCESS);
}


seccomp/seccomp_multiarch.c
/*************************************************************************\
*                  Copyright (C) Michael Kerrisk, 2017.                   *
*                                                                         *
* This program is free software. You may use, modify, and redistribute it *
* under the terms of the GNU General Public License as published by the   *
* Free Software Foundation, either version 3 or (at your option) any      *
* later version. This program is distributed without any warranty.  See   *
* the file COPYING.gpl-v3 for details.                                    *
\*************************************************************************/

/* seccomp_multiarch.c

   This program demonstrates one of the reasons that a seccomp filter must
   verify the architecture on *each* system call. The point is that some
   hardware platforms support multiple architectures. For example, modern
   x86 hardware supports x86-64, i386, and x32 architectures. The
   architecture is determined by the binary that is executed, and since a
   process may execute multiple binaries over its lifetime (via successive
   calls to execve()), this means that the architecture may change during
   the life of the process. However, once installed, seccomp filters
   persist for the lifetime of the process; in particular, filters persist
   across calls to execve() (if the filters permit calls to execve()), and
   thus across architecture changes.  Thus, a seccomp filter needs to check
   its architecture assumption each time the filter is invoked.

   The idea behind this program is to build binaries for different
   architectures from this source file. A chain of execve() calls inside
   one process can then be triggered by running one of the binaries with
   the other binary/binaries as its command-line argument(s). The first
   binary that is executed installs a seccomp() filter which remains
   installed for the life of the process. This filter causes calls to
   mq_notify(2) to fail, but checks for each of the syscall numbers for
   x86-64, i386, and x32, and fails each different architecture-specific
   with a differient error. An environment variable is employed to ensure
   tht only the first binary installs a seccomp() filter.

   Each binary prints out some status information, calls mq_notify() (and
   prints the resulting error number returned in 'errno'), and execs the
   next binary with the remaining command-line arguments.

   Demonstration steps:

   $ cc      -o seccomp_multiarch      seccomp_multiarch.c -lrt
   $ cc -m32 -o seccomp_multiarch_i386 seccomp_multiarch.c -lrt
   $ ./seccomp_multiarch ./seccomp_multiarch_i386 ./seccomp_multiarch

   NOTE: This program is designed to be run on an Intel x86-64 system.
   If you run this program on a non-Intel multi-arch platform, the
   definitions of architecture and system call numbers below will
   need to be adjusted.
*/
#define _GNU_SOURCE
#include <mqueue.h>
#include <stddef.h>
#include <fcntl.h>
#include <linux/audit.h>
#include <sys/syscall.h>
#include <linux/filter.h>
#include <linux/seccomp.h>
#include <sys/prctl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>

#define errExit(msg)    do { perror(msg); exit(EXIT_FAILURE); \
                        } while (0)

static int
seccomp(unsigned int operation, unsigned int flags, void *args)
{
    return syscall(__NR_seccomp, operation, flags, args);
}

/* Define names for the various system call numbers for the mq_notify()
   system call on Intel architectures:

   x86-64: [cc ...]
        This is the modern 64-bit Intel architecture with 64-bit
        instruction set and 64-bit pointers.

   i386: [cc -m32 ...]
        This is the historical Intel architecture with 32-bit instructions
        and 32-bit pointers (and thus 32-bit address space). As a backward
        compatibility feature, modern 64-bit Intel CPUs can also execute
        i386 programs.

   x32: [cc -mx32 ...]
        This is an alternate ABI for the Intel 64-bit architecture, available
        via the kernel configuration option CONFIG_X86_X32.  This ABI, which
        was added in Linux 3.4 (2012), allows an application to take advantage
        of the modern x86-64 instruction set and 64-bit registers (which are
        more numerous than the register set on i386) while at the same time
        employing 32-bit pointers (and thus 32-bit address space) and 32-bit
        'long'.  The rationale for adding the ABI was that the smaller types
        can result in significant memory savings in some code, and can also
        provide performance improvements in some cases.

        Because some types (e.g., pointers and 'long') under this ABI have
        sizes that are different from the x86-64 ABI, some system calls
        require signatures that differ from x86-64, and thus have different
        syscall numbers.  These numbers all have values >= 512 (see the
        kernel source file arch/x86/entry/syscalls/syscall_64.tbl), which
        allows some headroom to add new system call numbers for the base
        x86-64 architecture (for example, as at Linux 4.16, the highest
        x86-64 system call number 332 is for statx()).  In addition, when an
        x32 binary makes makes any system call, bit 30 (0x40000000) is set
        in the system call number, which enables the kernel code to
        implement 32-bit "compatibility mode" in cases where it is needed.

        In order to build programs for the x32 architecture, one must have
        user-space support (i.e., binutils tools built for x32 and libraries
        such as glibc built to this ABI), and in many distributions this
        support is *not* provided. A notable distribution that does support
        x32 is Debian (https://wiki.debian.org/X32Port).

        For further information on the x32 ABI, see
        https://sites.google.com/site/x32abi/ and
        https://lwn.net/Articles/456731/
*/

/* For the x32 ABI, all system call numbers have bit 30 set */

#define X32_SYSCALL_BIT         0x40000000

#define NR_mq_notify_x86_64     244
#define NR_mq_notify_i386       281
#define NR_mq_notify_x32        527     /* Actual syscall will also have
                                           X32_SYSCALL_BIT bit set */

static void
install_filter(void)
{
    struct sock_filter filter[] = {

        /* Load architecture into accumulator */

        BPF_STMT(BPF_LD | BPF_W | BPF_ABS,
                (offsetof(struct seccomp_data, arch))),

        /* Kill the process if the architecture is not one of those
           that we expect */

        BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, AUDIT_ARCH_X86_64, 2, 0),
        BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, AUDIT_ARCH_I386, 1, 0),
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_KILL),

        /* Load system call number */

        BPF_STMT(BPF_LD | BPF_W | BPF_ABS,
                 (offsetof(struct seccomp_data, nr))),

        /* Cause mq_notify() to fail with a different error on each
           architecture. (mq_notify() is an example of a system call that
           has a different number on each of x86-64, i386, and x32.)
           To make it easy to determine which system call number triggered
           the seccomp failure, return the system call number as the error
           number that will appear in 'errno'. */

        BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, NR_mq_notify_x86_64, 0, 1),
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ERRNO | NR_mq_notify_x86_64),

        BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K, NR_mq_notify_i386, 0, 1),
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ERRNO | NR_mq_notify_i386),

        BPF_JUMP(BPF_JMP | BPF_JEQ | BPF_K,
                 (NR_mq_notify_x32 + X32_SYSCALL_BIT), 0, 1),
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ERRNO | NR_mq_notify_x32),
                /* Return NR_mq_notify_x32, rather than
                   (NR_mq_notify_x32 + X32_SYSCALL_BIT) because glibc only
                   considers negative syscall return values in the range
                   [-4096 < retval < 0] to be error values that should be
                   (negated and) copied to 'errno'.  Note also that even
                   allowing for the masking of the X32_SYSCALL_BIT,
                   mq_notify() is an example of a system call that has a
                   different number on x86-64 and x32 (this is *not* the
                   case for all syscalls under these two ABIs). */

        /* All other system calls are allowed */

        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),

    };

    struct sock_fprog prog = {
        .len = (unsigned short) (sizeof(filter) / sizeof(filter[0])),
        .filter = filter,
    };

    if (seccomp(SECCOMP_SET_MODE_FILTER, 0, &prog) == -1)
        errExit("seccomp");
}

int
main(int argc, char **argv)
{
    const char *no_seccomp_filter_var = "NO_SECCOMP_FILTER";

    /* If the NO_SECCOMP_FILTER environment variable is not set, this should
       be the first invocation of the program, so install the seccomp filter */

    if (getenv(no_seccomp_filter_var) == NULL) {
        if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0))
            errExit("prctl");

        printf("Initial execution; installing seccomp filter\n");
        install_filter();
    }

    /* Print some information about the execution environment */

    printf("\targv[0] = %s; PID = %ld\n", argv[0], (long) getpid());
    char buf[1024];
    snprintf(buf, sizeof(buf), "echo -n '\t'; grep Seccomp /proc/%ld/status",
            (long) getpid());
    system(buf);
    snprintf(buf, sizeof(buf),
            "echo -n '\t'; file -L /proc/%ld/exe | sed 's/,.*//'",
            (long) getpid());
    system(buf);

    /* Call mq_notify() and check the result, which should be failure
       triggered by the seccomp filter. (The EBADF error is for the case
       where the seccomp filter did not fail the system call, and the
       kernel then returned an error because 0 is not a valid message
       queue descriptor.) */

    if (mq_notify(0, NULL) == -1 && errno != EBADF) {
        printf("mq_notify() failed with errno = %d\n", errno);
    } else {
        printf("mq_notify() was not failed by seccomp() filter!\n");
        printf("*** Make sure that the %s environment variable is not set\n",
                no_seccomp_filter_var);
        printf("    when running this program.\n");
        printf("*** Otherwise, check the syscall numbers in kernel headers\n");
        printf("    to see that they are correct\n");
        exit(EXIT_FAILURE);
    }

    /* If there is an argv[1], then execute the program named in argv[1].
       Beforehand, set the NO_SECCOMP_FILTER environment variable to tell
       the next program not to add a (duplicate) seccomp filter */

    if (argc > 1) {
        setenv(no_seccomp_filter_var, "y", 1);

        printf("\n");
        printf("About to exec: %s\n", argv[1]);
        execv(argv[1], &argv[1]);
        errExit("execvp");
    }

    exit(EXIT_SUCCESS);
}
